browser/ai_chat/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,354 @@
+  ai_chat_api_ =
+      std::make_unique<AIChatAPI>(web_contents->GetBrowserContext()
+                                      ->GetDefaultStoragePartition()
+                                      ->GetURLLoaderFactoryForBrowserProcess());
+}
+
+AIChatTabHelper::~AIChatTabHelper() = default;
+
+const std::vector<ConversationTurn>& AIChatTabHelper::GetConversationHistory() {
+  return chat_history_;
+}
+
+const std::string& AIChatTabHelper::GetConversationHistoryString() {
+  std::vector<std::string> turn_strings;
+  for (const ConversationTurn& turn : chat_history_) {
+    turn_strings.push_back((turn.character_type == CharacterType::HUMAN
+                                ? ai_chat::kHumanPrompt
+                                : ai_chat::kAIPrompt) +
+                           turn.text);
+  }
+
+  history_text_ = base::JoinString(turn_strings, "\n");
+
+  return history_text_;
+}
+
+void AIChatTabHelper::AddToConversationHistory(const ConversationTurn& turn) {
+  chat_history_.push_back(turn);
+
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::SetArticleSummaryString(const std::string& text) {
+  article_summary_ = text;
+}
+
+void AIChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+    obs.OnAPIRequestInProgress(IsRequestInProgress());
+  }
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::RequestSummary() {
+  if (!article_summary_.empty()) {
+    VLOG(1) << __func__ << " Article summary is in cache\n";
+
+    AddToConversationHistory({
+        CharacterType::ASSISTANT,
+        ConversationTurnVisibility::VISIBLE,
+        article_summary_,
+    });
+    return;
+  }
+
+  auto* primary_rfh = web_contents()->GetPrimaryMainFrame();
+
+  if (!primary_rfh) {
+    // TODO(petemill): Don't allow the UI to submit requests at this state
+    LOG(ERROR) << "Summary request submitted for a WebContents without a "
+                  "primary main frame";
+    for (auto& obs : observers_) {
+      obs.OnRequestSummaryFailed();
+    }
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  if (!rfh) {
+    // TODO(petemill): Don't allow the UI to submit requests at this state
+    LOG(ERROR) << "Summary request submitted for a WebContents without a"
+                  "primary AXTree-associated RenderFrameHost yet";
+    for (auto& obs : observers_) {
+      obs.OnRequestSummaryFailed();
+    }
+    return;
+  }
+
+  // TODO(@nullhook): Add a timeout and test this on real pages
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&AIChatTabHelper::OnSnapshotFinished,
+                     base::Unretained(this)),
```

Comment:
I think this is safe because the rfh should be destroyed before the webcontents, but callers outside of rfh/webcontents internals use WeakPtr so I think it's better to be safe here